### PR TITLE
PR for hyperband-enabled decision trees

### PR DESCRIPTION
**Pull request by RuohanW**
*Issue #, if available:* None

*Description of changes:*
This change enables multi-fidelity scheduler (HyperbandScheduler) to be used with decision tree-based methods (Catboost, lightGBM, Random Forests, and Extreme Trees), during hyperparameter optimization.

With the change, the decision tree-based methods is now able to perform incremental training, similar to those found in scikit-learn. By specifying warm_start=True in fit() function, the training would continue from the model trained over previous calls to fit().

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-785-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-785/2/index.html

**Comment by szha** [comment]
Job PR-785-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-785/3/index.html

File: tabular/src/autogluon/tabular/models/abstract/abstract_model.py [hunk 1]
```diff
@@ -106,6 +106,10 @@ def __init__(self, path: str, name: str, problem_type: str, eval_metric: Union[s
             self.nondefault_params = list(hyperparameters.keys())[:]  # These are hyperparameters that user has specified.
         self.params_trained = dict()
 
+        #for warm-starting with hyperband scheduler
```
Comment (mseeger) [inline tabular/src/autogluon/tabular/models/abstract/abstract_model.py:109]: Could you document somewhere what these variables are?

File: tabular/src/autogluon/tabular/models/abstract/abstract_model.py [hunk 2]
```diff
@@ -579,15 +583,24 @@ def hyperparameter_tune(self, X_train, y_train, X_val, y_val, scheduler_options,
                 if isinstance(params_copy[hyperparam], Space):
                     logger.log(15, f"{hyperparam}:   {params_copy[hyperparam]}")
 
+        epochs = kwargs.get('epochs', 1)
```
Comment (mseeger) [inline tabular/src/autogluon/tabular/models/abstract/abstract_model.py:586]: Not sure I like the name 'epochs'. This has a pretty specific meaning for NN models, and this here is an abstract model class.

File: tabular/src/autogluon/tabular/models/abstract/abstract_model.py [hunk 3]
```diff
@@ -579,15 +583,24 @@ def hyperparameter_tune(self, X_train, y_train, X_val, y_val, scheduler_options,
...
+
         util_args = dict(
             dataset_train_filename=dataset_train_filename,
             dataset_val_filename=dataset_val_filename,
             directory=directory,
             model=self,
             time_start=time_start,
             time_limit=scheduler_options['time_out'],
+            epochs=epochs,
+            report_probs=kwargs.get('report_probs', False)
```
Comment (mseeger) [inline tabular/src/autogluon/tabular/models/abstract/abstract_model.py:596]: What is report_probs doing?

File: tabular/src/autogluon/tabular/models/abstract/abstract_model.py [hunk 4]
```diff
@@ -601,7 +614,10 @@ def hyperparameter_tune(self, X_train, y_train, X_val, y_val, scheduler_options,
         scheduler.run()
         scheduler.join_jobs()
 
-        return self._get_hpo_results(scheduler=scheduler, scheduler_options=scheduler_options, time_start=time_start)
+        hpo_models, hpo_model_performances, hpo_results = self._get_hpo_results(scheduler=scheduler, scheduler_options=scheduler_options, time_start=time_start)
+        if 'report_probs' in kwargs:
+            hpo_results['ensemble'] = scheduler.ensemble
```
Comment (mseeger) [inline tabular/src/autogluon/tabular/models/abstract/abstract_model.py:619]: You are assuming here that if report_probs is given, the scheduler is of a very specific type, so it has ensemble member. This is a little reckless, at least the whole thing has to be documented somewhere.

File: tabular/src/autogluon/tabular/models/abstract/model_trial.py [hunk 5]
```diff
@@ -24,33 +25,40 @@ def model_trial(args, reporter: LocalStatusReporter):
...
+
         X_train, y_train = load_pkl.load(util_args.directory + util_args.dataset_train_filename)
         X_val, y_val = load_pkl.load(util_args.directory + util_args.dataset_val_filename)
 
         fit_model_args = dict(X_train=X_train, y_train=y_train, X_val=X_val, y_val=y_val)
         predict_proba_args = dict(X=X_val)
-        model = fit_and_save_model(model=model, params=args, fit_args=fit_model_args, predict_proba_args=predict_proba_args, y_val=y_val,
-                                   time_start=util_args.time_start, time_limit=util_args.get('time_limit', None), reporter=None)
+        model = fit_and_save_model(model=model, params=args, fit_args=fit_model_args, predict_proba_args=predict_proba_args,
+                                   y_val=y_val, time_start=util_args.time_start, time_limit=util_args.get('time_limit', None),
```
Comment (mseeger) [inline tabular/src/autogluon/tabular/models/abstract/model_trial.py:38]: .get('time_limit') instead of .get('time_limit', None)

Comment (mseeger) [reply tabular/src/autogluon/tabular/models/abstract/model_trial.py:38]: And why was reporter=None before?

File: tabular/src/autogluon/tabular/models/abstract/model_trial.py [hunk 6]
```diff
@@ -24,33 +25,40 @@ def model_trial(args, reporter: LocalStatusReporter):
...
         reporter.terminate()
-    else:
-        reporter(epoch=1, validation_performance=model.val_score)
 
 
 def prepare_inputs(args):
     task_id = args.pop('task_id')
     util_args = args.pop('util_args')
 
+    args.pop("epochs")
```
Comment (mseeger) [inline tabular/src/autogluon/tabular/models/abstract/model_trial.py:52]: Not your code, but I find all this popping and renaming args -> util_args super confusing

File: tabular/src/autogluon/tabular/models/abstract/model_trial.py [hunk 7]
```diff
@@ -24,33 +25,40 @@ def model_trial(args, reporter: LocalStatusReporter):
...
+
     file_prefix = f"trial_{task_id}"  # append to all file names created during this trial. Do NOT change!
     model = util_args.model  # the model object must be passed into model_trial() here
     model.name = model.name + os.path.sep + file_prefix
     model.set_contexts(path_context=model.path_root + model.name + os.path.sep)
     return model, args, util_args
 
 
-def fit_and_save_model(model, params, fit_args, predict_proba_args, y_val, time_start, time_limit=None, reporter=None):
+def fit_and_save_model(model, params, fit_args, predict_proba_args, y_val, time_start, time_limit=None, epochs=1, reporter=None, report_probs=False):
```
Comment (mseeger) [inline tabular/src/autogluon/tabular/models/abstract/model_trial.py:61]: In general, I prefer default values =None, and deal with it in the function, so epochs=None, report_probs=None

File: tabular/src/autogluon/tabular/models/abstract/model_trial.py [hunk 8]
```diff
@@ -62,13 +70,21 @@ def fit_and_save_model(model, params, fit_args, predict_proba_args, y_val, time_
...
+        time_fit_end = time.time()
+        y_pred_proba = model.predict_proba(**predict_proba_args)
+        time_pred_end = time.time()
+        val_score = model.score_with_y_pred_proba(y=y_val, y_pred_proba=y_pred_proba)
+        model.val_score = val_score
+        model.fit_time = time_fit_end - time_fit_start
+        model.predict_time = time_pred_end - time_fit_end
+        if report_probs:
+            #use model.path as a unique identifier. maybe change this later?
+            reporter(epoch=i+1, validation_performance=val_score, y_probs=y_pred_proba, y_val=y_val, model_path=model.path)
```
Comment (mseeger) [inline tabular/src/autogluon/tabular/models/abstract/model_trial.py:85]: Have you thought about storing the prediction vectors here, instead of sending them back to the scheduler?

Comment (mseeger) [reply tabular/src/autogluon/tabular/models/abstract/model_trial.py:85]: I mean, you store the model here, why not the prediction vector as well?

File: tabular/tests/unittests/test_hyperband_for_decision_trees.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+    scheduler_options = compile_scheduler_options(
+        scheduler_options=scheduler_options,
+        search_strategy=strategy,
+        search_options=None,
+        nthreads_per_trial=thread_per_trial,
+        ngpus_per_trial=0,
+        checkpoint=None,
+        num_trials=num_trials,
+        time_out=timeout, #1 hours timeout
```
Comment (mseeger) [inline tabular/tests/unittests/test_hyperband_for_decision_trees.py:74]: Remove comment

File: tabular/src/autogluon/tabular/models/lgb/hyperparameters/lgb_trial.py [hunk 10]
```diff
@@ -18,21 +22,33 @@ def lgb_trial(args, reporter):
...
-                                       time_start=util_args.time_start, time_limit=util_args.get('time_limit', None), reporter=reporter)
+                                       time_start=util_args.time_start, time_limit=util_args.get('time_limit', None),
+                                       epochs=epochs, reporter=reporter, report_probs=report_probs)
     except Exception as e:
-        if not isinstance(e, TimeLimitExceeded):
+        if not (isinstance(e, TimeLimitExceeded) or isinstance(e, AutoGluonEarlyStop)):
             logger.exception(e, exc_info=True)
         reporter.terminate()
+        if isinstance(e, AutoGluonEarlyStop):
+            model.save()
```
Comment (gradientsky) [inline tabular/src/autogluon/tabular/models/lgb/hyperparameters/lgb_trial.py:51]: ditto on exceptions handling

File: tabular/src/autogluon/tabular/models/rf/rf_model.py [hunk 11]
```diff
@@ -55,62 +57,85 @@ def _set_default_params(self):
...
-        if n_estimators_final > n_estimators_test * 2:
+                n_estimators_test = 8
+
+            n_estimator_increments = [n_estimators_final]
+
+            # Very rough guess to size of a single tree before training
             if self.problem_type == MULTICLASS:
-                n_estimator_increments = [n_estimators_test, n_estimators_final]
-                hyperparams['warm_start'] = True
+                if self.num_classes is None:
```
Comment (gradientsky) [inline tabular/src/autogluon/tabular/models/rf/rf_model.py:101]: Is this possible? We always should know how many classes are in the model.

File: tabular/src/autogluon/tabular/models/rf/rf_model.py [hunk 12]
```diff
@@ -55,62 +57,85 @@ def _set_default_params(self):
...
-        bytes_per_estimator = num_trees_per_estimator * len(X_train) / 60000 * 1e6  # Underestimates by 3x on ExtraTrees
-        available_mem = psutil.virtual_memory().available
-        expected_memory_usage = bytes_per_estimator * n_estimators_final / available_mem
-        expected_min_memory_usage = bytes_per_estimator * n_estimators_minimum / available_mem
-        if expected_min_memory_usage > (0.5 * max_memory_usage_ratio):  # if minimum estimated size is greater than 50% memory
-            logger.warning(f'\tWarning: Model is expected to require {round(expected_min_memory_usage * 100, 2)}% of available memory (Estimated before training)...')
-            raise NotEnoughMemoryError
-
-        if n_estimators_final > n_estimators_test * 2:
+                n_estimators_test = 8
```
Comment (gradientsky) [inline tabular/src/autogluon/tabular/models/rf/rf_model.py:95]: Not sure if these estimators are good picks for any dataset. Ideally this should be a function of dataset length and columns number.

**Review by gradientsky — CHANGES_REQUESTED** [review]

File: tabular/src/autogluon/tabular/models/abstract/abstract_model.py (hunk 1 above)
Comment (RuohanW) [reply tabular/src/autogluon/tabular/models/abstract/abstract_model.py:109]: The variables are updated with comments.

File: tabular/src/autogluon/tabular/models/abstract/abstract_model.py (hunk 2 above)
Comment (RuohanW) [reply tabular/src/autogluon/tabular/models/abstract/abstract_model.py:586]: I can think of names like "rounds" or "train_iters". Can we decide one first before I update the code? This would require updates at multiple places. So I'd like the name to be fixed first. Epoch is actually quite close to the actual meaning, since we iterate this number of times to add trees to different base learners.

File: tabular/src/autogluon/tabular/models/abstract/abstract_model.py (hunk 3 above)
Comment (RuohanW) [reply tabular/src/autogluon/tabular/models/abstract/abstract_model.py:596]: This is a flag for whether validation set probabilities need to be passed to the scheduler. This is needed for the hpo over ensemble.

File: tabular/src/autogluon/tabular/models/abstract/abstract_model.py (hunk 4 above)
Comment (RuohanW) [reply tabular/src/autogluon/tabular/models/abstract/abstract_model.py:619]: This is also for hpo over ensemble. report_probs is a flag, it doesn't actually contain any probability data. I have added another check hasattr(scheduler, "ensemble") to be safe. Also added the comment to document this change.

File: tabular/src/autogluon/tabular/models/abstract/model_trial.py (hunk 5 above)
Comment (RuohanW) [reply tabular/src/autogluon/tabular/models/abstract/model_trial.py:38]: It was none because for other non-hyperband schedulers, there was no need to report back anything during the training. Updated the get("time_limit")

File: tabular/src/autogluon/tabular/models/abstract/model_trial.py (hunk 7 above)
Comment (RuohanW) [reply tabular/src/autogluon/tabular/models/abstract/model_trial.py:61]: code updated

File: tabular/tests/unittests/test_hyperband_for_decision_trees.py (hunk 9 above)
Comment (RuohanW) [reply tabular/tests/unittests/test_hyperband_for_decision_trees.py:74]: comments removed

File: tabular/src/autogluon/tabular/models/abstract/model_trial.py (hunk 8 above)
Comment (RuohanW) [reply tabular/src/autogluon/tabular/models/abstract/model_trial.py:85]: Will rework the validation probability passing

File: tabular/src/autogluon/tabular/models/rf/rf_model.py (hunk 12 above)
Comment (RuohanW) [reply tabular/src/autogluon/tabular/models/rf/rf_model.py:95]: The code was refactoring the previous implementation. I didn't change any heuristics that existed before. I think any new changes to RF model should be handled by the AG team.

**Comment by szha** [comment]
Job PR-785-9 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-785/9/index.html

**Comment by Innixma** [comment]
@RuohanW The git diff for this commit seems to be corrupted, as many of these changes are from other PRs that were merged into master. You may want to rebase your PR against master or submit a new PR to avoid the corruption, otherwise it is very hard to determine which changes are yours and which are from other PRs that have already been merged.

As a general rule, I would recommend always using rebase when changes occur in master to keep things simple.
